### PR TITLE
✨ [Network/Client] Implement client input packet system with command-line parsing and connection handling

### LINKED ISSUE #29: Implement Client Input Packet (Mock)
# Implement Client Input Packet (Mock)

**Milestone:** Milestone 2  
**Labels:** area/network, area/client, type/feature, P1

## Description
Send mock input packets (keyboard state) from client to server.

## Tasks
- Capture input state
- Create mock input binary packet
- Send packet every tick

## Acceptance Criteria
- Client logs confirm packets are sent
- Server receives mock input packets

### PR DESCRIPTION
**Pull request by Enoal-Fauchille-Bolle**
## 🚀 Pull Request Summary

Closes: #29
Milestone: 2 - Engine Core, Networking Base Client Prototype

---

## 🧩 Type of Change

Select all relevant options:

- [x] ✨ Feature
- [ ] 🐛 Bug Fix
- [x] ♻️ Refactor
- [ ] 📝 Documentation
- [x] 🧪 Tests
- [ ] 🔧 Tooling / CI
- [ ] 🚀 Performance
- [ ] Other (please describe)

---

## 📦 Changes Included

List the main changes made in this PR:

- Implemented command-line argument parser for client (server IP and port)
- Added RFC-compliant input-to-bitfield conversion system for network packets
- Integrated network components (NetworkSender, NetworkReceiver, NetworkInputPacket) into GameWorld initialization
- Implemented TCP connection polling with timeout handling and improved error management
- Created comprehensive unit tests for InputToBitfield, command-line parser, and network functionality
- Refactored test infrastructure with ArgvHelper utility and shared pointer state management

---

## 🧪 How to Test

Describe how reviewers can test your changes:

1. Build the project (`cmake --build build`)
2. Start the server: `./build/server/r-type_server 50000`
3. Start the client with server address: `./build/client/r-type_client 127.0.0.1 50000 Player`
4. Use keyboard inputs (arrow keys, space) to verify input packet transmission
5. Check server logs for received input packets

**Test Commands:**
```bash
# Run all unit tests
cd build && ctest --output-on-failure

# Run specific input tests
./build/tests/test_input_to_bitfield
./build/tests/test_command_line_parser
./build/tests/test_network
```

**Expected Results:**
- Client successfully parses command-line arguments
- Client establishes TCP connection to server
- Input state is converted to bitfield format
- Network packets are sent from client to server
- Server receives and logs input packets

---

## 🗂️ Related Areas

Which part of the project does this PR affect?

- [ ] area/docs
- [ ] area/server
- [x] area/client
- [ ] area/engine
- [x] area/tests

---

## 🔍 Checklist Before Requesting Review

Please confirm the following items are done:

- [x] My commits follow the **Gitmoji + English commit message** convention
- [x] My branch name follows the required format (`feature/29-implement-client-input-packet-mock`)
- [x] I have linked the PR to an Issue (`Closes #29`)
- [x] I ran and passed **all Git hooks**
- [x] I added or updated documentation if needed
- [x] I added or updated tests if needed
- [x] The code builds without errors
- [x] I performed manual testing
- [x] CI passes all checks (build + tests + formatting)

---

## 📸 Screenshots / Logs (if applicable)

*Client successfully connecting and sending input packets to server*

---

## 👀 Reviewer Notes

Anything you want reviewers to pay special attention to?

- **Network Architecture**: Review the integration of NetworkSender/NetworkReceiver components into GameWorld
- **Input System**: Verify RFC-compliant bitfield conversion matches protocol specification
- **Error Handling**: Check TCP connection timeout and polling implementation
- **Test Coverage**: Ensure test refactoring with shared pointers properly manages state
- **Command-Line Parsing**: Validate argument validation and error handling for edge cases

### CODE REVIEW

File: client/engine/systems/systems_functions/NetworkInputSystem.cpp [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+                      << " Down=" << ((input_bitfield & (1 << 1)) ? "1" : "0")
+                      << " Left=" << ((input_bitfield & (1 << 2)) ? "1" : "0")
+                      << " Right=" << ((input_bitfield & (1 << 3)) ? "1" : "0")
+                      << " Shoot=" << ((input_bitfield & (1 << 4)) ? "1" : "0")
+                      << ")" << std::endl;
+            last_sent_input_ = input_bitfield;
+        }
+    }
+}
+}  // namespace Rtype::Client
```
[the remark this answers is not in the document]
Comment (Enoal-Fauchille-Bolle) [reply client/engine/systems/systems_functions/NetworkInputSystem.cpp]: Fixed in https://github.com/MatthieuGA/R_TYPE_J.A.M.E.S./pull/134/commits/a633266d448f4eef65703d7c49214410532b3c1c

File: tests/test_network.cpp [hunk 2]
```diff
@@ -223,25 +230,34 @@ TEST(NetworkTest, ConnectToServerSendsConnectReqPacket) {
...
-                EXPECT_FALSE(ec);
-                connection_complete = true;
-            });
+            auto ack = std::make_shared<std::vector<uint8_t>>(
+                BuildConnectAckPacket(5, 0));
+            server.AsyncWrite(*ack,
+                [connection_complete, ack](boost::system::error_code ec) {
+                    if (ec)
+                        return;
+                    EXPECT_FALSE(ec);
```
[the remark this answers is not in the document]
Comment (Enoal-Fauchille-Bolle) [reply tests/test_network.cpp]: Fixed in https://github.com/MatthieuGA/R_TYPE_J.A.M.E.S./pull/134/commits/a7059934a8fdeb0f3e23ae94a544ce5f56d67753

File: tests/test_network.cpp [hunk 3]
```diff
@@ -223,25 +230,34 @@ TEST(NetworkTest, ConnectToServerSendsConnectReqPacket) {
...
+            return;
         EXPECT_FALSE(ec);
-        std::vector<uint8_t> req(44);
-        server.AsyncRead(req, [&](boost::system::error_code ec, size_t) {
+        auto req = std::make_shared<std::vector<uint8_t>>(44);
+        server.AsyncRead(*req, [&, connection_complete, req](
+                                   boost::system::error_code ec, size_t) {
+            if (ec)
+                return;
             EXPECT_FALSE(ec);
```
[the remark this answers is not in the document]
Comment (Enoal-Fauchille-Bolle) [reply tests/test_network.cpp]: Fixed in https://github.com/MatthieuGA/R_TYPE_J.A.M.E.S./pull/134/commits/a7059934a8fdeb0f3e23ae94a544ce5f56d67753

File: tests/test_network.cpp [hunk 4]
```diff
@@ -193,27 +194,33 @@ TEST(NetworkTest, DisconnectBeforeConnectDoesNotThrow) {
...
+    auto received_data = std::make_shared<std::vector<uint8_t>>(
         44);  // 12-byte header + 32-byte username
 
     MockTcpServer server(io, 4242);
-    server.AsyncAccept([&](boost::system::error_code ec) {
+    server.AsyncAccept([&, connect_req_received, received_data](
+                           boost::system::error_code ec) {
+        if (ec)
+            return;
         EXPECT_FALSE(ec);
```
[the remark this answers is not in the document]
Comment (Enoal-Fauchille-Bolle) [reply tests/test_network.cpp]: Fixed in https://github.com/MatthieuGA/R_TYPE_J.A.M.E.S./pull/134/commits/a7059934a8fdeb0f3e23ae94a544ce5f56d67753
